cEP-0025.md

```diff
@@ -0,0 +1,116 @@
+# Integration of pyflakes-enhanced-AST into coala
+
+| Metadata |                                                 |
+| -------- | ----------------------------------------------- |
+| cEP      | 22                                              |
+| Version  | 1.0                                             |
+| Title    | Integration of pyflakes-enhanced-AST into coala |
+| Authors  | Ankit Joshi <mailto:ajankit2304@gmail.com>      |
+| Status   | Proposed                                        |
+| Type     | Feature                                         |
+
+## Abstract
+
+This document describes how to create a metabear to integrate
+pyflakes-enhanced-AST into coala.
+
+## Introduction
+
+The current approach of plugin development for flake8 does not utilises
+the full potential of enhanced AST rather they simply depend on python AST
+and so, a whole lot of rework is required to do the basic traversing
+and collection of important nodes.
+
+Pyflakes provides with a basic API that does the traversing.
+So, if a developer uses enhanced AST he just needs to work on
+the implementation of the new logic that his/her plugin provides
+and not about the fidelity of the basic node handlers.
+
+## Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. There will be a separate repository named as coala-pyflakes which will
+   be installable via `pip install coala-pyflakes`.
+2. The repository will contain a new package `pyflakes-plugins` which
+   would house all bears which uses pyflakes-enhanced-ast.
+3. The `pyflakes-plugins` would contain a metabear `PyflakesASTBear`
+   which would be used by all other plugin bears.
+
+## Management of the new `coala-pyflakes` repository
+
+The repository will use a similar mechanism of what is being
+used in `coala-bears` to test all its plugins.
+
+## Code Samples/Prototypes
+
+Here is a prototype for the implementations of PyflakesASTBear:
+
+```python
+# Imports not mentioned to maintain brevity
+
+class PyflakesResult(HiddenResult):
+
+    @enforce_signature
+    def __init__(self, origin, deadScopes):
+
+        Result.__init__(self, origin, message='')
+
+        self.class_scope = self.get_scope(ClassScope, deadScopes)
+        self.function_scope = self.get_scope(FunctionScope, deadScopes)
+        self.module_scope = self.get_scope(ModuleScope, deadScopes)
+        self.generator_scope = self.get_scope(GeneratorScope, deadScopes)
+        self.doctest_scope = self.get_scope(DoctestScope, deadScopes)
+
+    def get_scope(self, scope_level, scopes):
```

Comment:
get_scope -> get_scopes ; always use plural to indicate a list is returned.

and then update the instance attributes.